### DISCUSSION TITLE
How to call TestClient with requests package ?

### DISCUSSION THREAD

**Discussion by thomassajot** [fastapi/fastapi#8462] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from typing import Dict, Optional
from pydantic import BaseModel, Field
import requests
from fastapi import FastAPI, Request
from fastapi.testclient import TestClient


class BackendResponse(BaseModel):
    data: Optional[Dict] = Field(None, title="Primary response data.")
    
class MockBackendServer(FastAPI):

    def __init__(self):
        super().__init__(title='mock', version='0.0.1')

        @self.get('/{fn}', response_model=BackendResponse)
        def mock_backend(request: Request, fn: str) -> BackendResponse:
            return BackendResponse(data={'output': 'mock', 'function': fn})

server = MockBackendServer()
client = TestClient(server)
print(client.base_url)
requests.get(client.base_url + '/hello')
# requests.exceptions.ConnectionError: HTTPConnectionPool(host='testserver', port=80): Max retries exceeded with url: /hello (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fd4da9b0220>: Failed to establish a new connection: [Errno 8] nodename nor servname provided, or not known'))
```

### Description

I would like to use standard `requests.get` request to query a TestClient server.
However I get an error while doing so, not sure why.
How can I call a TestClient without using the `client.get` method ? 

### Operating System

macOS

### FastAPI Version

0.75.0

### Python Version

3.9

**Comment by Kludex** [comment]
`TestClient` uses `requests` underneath... Why do you want to call it directly?

To understand what's happening, you can check:
- [requests transport adapters](https://2.python-requests.org/en/master/user/advanced/#id15)
- [`TestClient` source code](https://github.com/encode/starlette/blob/master/starlette/testclient.py)

**Comment by thomassajot** [comment]
I have 2 services. The first one is a proxy, the second one is a dummy server (for test purposes). 
The proxy calls the second server using `requests.get`. If I spin up the second server using TestClient, then the call fails.

**Comment by thomassajot** [comment]
I already went through the source code of TestClient. I am not sure why I cannot replicate the call using vanilla `requests` package.

**Comment by accelleon** [comment, marked as answer]
As Kludex was trying to point towards when he linked requests' adapters:

Starlette's `TestClient` doesn't actually make a request to any domain, localhost or otherwise. What it actually does is communicate directly with the app's event handler for the backend, the requests are intercepted by the `_ASGIAdapter` defined in Starlette's `testclient.py` and it uses communication between threads (via anyio portals) to emulate the request.

That override happens when we override the adapter for the requests `Session` here:
```python
class TestClient(requests.Session):
...
    def __init__(
...
        adapter = _ASGIAdapter(
            asgi_app,
            portal_factory=self._portal_factory,
            raise_server_exceptions=raise_server_exceptions,
            root_path=root_path,
        )
        self.mount("http://", adapter)
        self.mount("https://", adapter)
        self.mount("ws://", adapter)
        self.mount("wss://", adapter)
```

You need to keep in mind that ASGI is 2 parts:
* A server that receives client connections, and communicates with the backend via events
* A backend that receives events from the server processes them and sends an event to the server to communicate the response to the client
(see https://asgi.readthedocs.io/en/latest/introduction.html)

If you want to set up the proxy to make requests to the test client, you're likely going to have to some *real* jank to talk to the threads spun up by another process. That would be well beyond the scope of a discussion on FastAPI. Honestly, if testing your proxy is your goal here, write a script that starts the backend server then runs `pytest` against your proxy.

If you're adverse to that idea, the more complicated thing to do is mock responses from your backend. You'd likely have to manually build several `Response` objects and intercept outbound requests from your proxy.
